### PR TITLE
Support configurable OpenID Connect Identity Provider

### LINKED ISSUE #181: Support existing Keycloak instance
Thanks for the OIDC plugin and including a setup for Keycloak in this helm chart! 

For those of us who already have existing Keycloak instances, it would be great to be able to leverage those by configuring the Keycloak component to point to our existing instance rather than deploying a new one.

### PR DESCRIPTION
**Pull request by devinbost**
Fixes #181

The test procedures for evaluating these changes are available in the updated readme.

This PR adds significant functionality for supporting Starlight for Kafka workloads with OIDC and OAuth2 for both TLS and non-TLS environments.

### CODE REVIEW

File: helm-chart-sources/pulsar/templates/function/function-configmap.yaml [hunk 1]
```diff
@@ -127,6 +127,9 @@ data:
       {{- if .Values.keycloak.enabled }}
       openIDAllowedTokenIssuers: "{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }},{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}.{{ .Release.Namespace }}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }},{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}.{{ template "pulsar.serviceDnsSuffix" . }}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }}"
       {{- end }}
+      {{- if .Values.openid.enabled }}
```
Comment (devinbost) [inline helm-chart-sources/pulsar/templates/function/function-configmap.yaml:130]: Not sure if I still need this?

File: helm-chart-sources/pulsar/values.yaml [hunk 2]
```diff
@@ -1752,6 +1752,29 @@ pulsarHeartbeat:
       topicName: "persistent://public/default/pubsub-latency-test"
   service: {}
 
+# For configuring OpenID in Okta for Pulsar, instructions are here: https://www.youtube.com/watch?v=UQBrecHOXxU&ab_channel=DataStaxDevelopers
+# Then, use the values from the Payload generated to populate these fields:
+openid:
+  enabled: false
+  allowedAudience: api://pulsarClient
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/values.yaml:1759]: What do you think about using `""` for the default here and then requiring a user to supply the audience? I might need to think about this a bit more.

**Review by michaeljmarshall — COMMENTED** [review]
Thank you for your contribution @devinbost! I left a lot of feedback. Most of it is pretty minor, but there are some that we'll definitely need to address before merging this. Please let me know if you have any questions.

File: examples/kafka/dev-values-kafka-proxy.yaml [hunk 3]
```diff
@@ -161,7 +161,7 @@ proxy:
 # The Proxy uses the same TLS certificate configured for the Pulsar protocol
 #    PULSAR_PREFIX_kafkaAdvertisedListeners: "SASL_PLAINTEXT://pulsar-proxy:9092,SASL_SSL://pulsar-proxy:9093"
 # if you use port forwarding to localhost then you have to set 'localhost'    
-    PULSAR_PREFIX_kafkaAdvertisedListeners: "SASL_PLAINTEXT://localhost:9092,SASL_SSL://localhost:9093"
+    PULSAR_PREFIX_kafkaAdvertisedListeners: "SASL_PLAINTEXT://pulsar-proxy:9092,SASL_SSL://pulsar-proxy:9093"
```
Comment (eolivelli) [inline examples/kafka/dev-values-kafka-proxy.yaml:164]: please do not change this example
this is supposed to work with port-forwarding

actually we could add two listeners, one that advertises itself on "localhost" and one on "pulsar-proxy", on 2 different ports, this way it will be accessible both via port-forwarding and from the k8s cluster
but I would keep this change as a separate PR

File: helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml [hunk 4]
```diff
@@ -83,15 +83,69 @@ data:
   tokenPublicKey: "file:///pulsar/token-public-key/{{ .Values.tokenPublicKeyFile }}"
   brokerClientAuthenticationPlugin: "org.apache.pulsar.client.impl.auth.AuthenticationToken"
   brokerClientAuthenticationParameters: "file:///pulsar/token-superuser/superuser.jwt"
-  {{- if .Values.brokerSts.extraAuthProvider }}
-  authenticationProviders: "{{ .Values.brokerSts.authenticationProviders }},{{ .Values.brokerSts.extraAuthProvider }}"
-  {{- else }}
-  authenticationProviders: "{{ .Values.brokerSts.authenticationProviders }}"
-  {{- end }}
```
[the remark this answers is not in the document]
Comment (devinbost) [reply helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml:90]: How do we converge this with the logic that simplifies configuration of enabling S4K?

File: helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml [hunk 5]
```diff
@@ -83,15 +83,69 @@ data:
...
+  # Use the following block if we want to also enable OAuth2 between Pulsar components, not just between clients and Pulsar:
+  #{{- if .Values.openid.enabled }}
+  #brokerClientAuthenticationPlugin: "org.apache.pulsar.client.impl.auth.oauth2.AuthenticationOAuth2"
+  #brokerClientAuthenticationParameters: "{\"privateKey\":\"file:///pulsar/openid-client-credentials/superuserKey.json\",\"issuerUrl\":\"{{ .Values.openid.issuerUrl }}\",\"scope\":\"{{ .Values.openid.scope }}\",\"audience\":\"{{ .Values.openid.allowedAudience }}\"}"
+  #{{- else }} # There's probably a smarter way to write this so it's more readable but gets trimmed to one line so Jackson doesn't blow up on new lines when parsing.
+  #brokerClientAuthenticationPlugin: "org.apache.pulsar.client.impl.auth.AuthenticationToken"
+  #brokerClientAuthenticationParameters: "file:///pulsar/token-superuser/superuser.jwt"
+  #{{- end }}
```
[the remark this answers is not in the document]
Comment (devinbost) [reply helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml:93]: I left this in here as more of a discussion point... I'm leaning towards needing a separate PR to allow us to enable OpenID between each server component, but I think we at least need a way to ensure we can add that functionality without needing to make major changes to the chart.

File: helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml [hunk 6]
```diff
@@ -83,15 +83,69 @@ data:
...
   {{- if .Values.keycloak.enabled }}
   PULSAR_PREFIX_openIDAllowedTokenIssuers: "{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }},{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}.{{ .Release.Namespace }}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }},{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}.{{ template "pulsar.serviceDnsSuffix" . }}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }}"
   {{- end }}
   {{- end }}
+
+# Putting this section down here so it's easier to read:
+{{- if .Values.openid.enabled }}
+  PULSAR_PREFIX_openIDAllowedTokenIssuers: "{{ .Values.openid.issuerUrl }}"
```
[the remark this answers is not in the document]
Comment (devinbost) [reply helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml:101]: Is there a way that we can handle both use cases without needing to branch the conditional logic in the chart?

File: helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml [hunk 7]
```diff
@@ -83,15 +83,68 @@ data:
...
+  # Java property configuration file of OauthValidatorCallbackHandler:
+  PULSAR_PREFIX_kopOauth2ConfigFile: "/pulsar/conf/kop-handler.properties"
+{{- end }}
+{{- if and .Values.openid.enabled .Values.openid.withS4k .Values.enableTls }}
+  PULSAR_PREFIX_kopTlsEnabledWithBroker: "true"
+  PULSAR_PREFIX_kafkaListeners: "SASL_PLAINTEXT://0.0.0.0:9092,SASL_SSL://0.0.0.0:9093" 
+  PULSAR_PREFIX_kafkaAdvertisedListeners: "SASL_PLAINTEXT://advertisedAddress:9092,SASL_SSL://advertisedAddress:9093" 
+{{- end }}
+{{- if and .Values.openid.enabled .Values.openid.withS4k (not .Values.enableTls ) }}
+  PULSAR_PREFIX_kopTlsEnabledWithBroker: "true"
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/broker-sts/broker-sts-configmap.yaml:144]: Should this be `false`?

File: examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+  websocket:
+    enableTlsWithBroker: true
+    enableHostnameVerification: true
+    certSpec:
+      privateKey:
+        algorithm: "ECDSA"
+        size: 256
+  proxy:
+    enableTlsWithBroker: true
+    enableHostnameVerification: false
```
Comment (devinbost) [inline examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml:113]: TO DO: Need a secure default

File: examples/kafka/demos/demo.sh [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+# To test OpenID/OAuth2 on non-TLS endpoint:
+# Make sure that the oauth parameters below map to your actual configs in the helm values file!
+cat > /pulsar/kafka/kafka_2.12-3.3.2/config/producer.properties
+bootstrap.servers=pulsar-proxy:9092
+compression.type=none
+sasl.login.callback.handler.class=com.datastax.oss.kafka.oauth.OauthLoginCallbackHandler
+security.protocol=SASL_PLAINTEXT
+sasl.mechanism=OAUTHBEARER
+sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule \
+   required oauth.issuer.url="https://dev-42506116.okta.com/oauth2/aus7ypk6sjvgF4l615d7"\
```
Comment (eolivelli) [inline examples/kafka/demos/demo.sh:71]: can we create a variable in the beginning of the .sh file with `dev-42506116.okta.com` ?

File: examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+  selfSignedPerComponent:
+    enabled: true
+
+openid:
+  enabled: true
+  # From token generated in Okta UI or other method:
+  issuerUrl: https://dev-42506116.okta.com/oauth2/aus7ypk6sjvgF4l615d7
+  scope: pulsar_client_m2m
+  client_id: "0oa7ypwvxnvo9xnDd5d7"
+  client_secret: "CL08ZNhF91fsCUm7rtYqHs-XUak5H7gLY01tF2bP"
```
Comment (eolivelli) [inline examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml:179]: we should not commit "secrets" on git

File: examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml (hunk 8 above)
Comment (michaeljmarshall) [reply examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml:113]: Let's update this before we merge.

File: helm-chart-sources/pulsar/values.yaml [hunk 11]
```diff
@@ -1752,6 +1752,31 @@ pulsarHeartbeat:
...
+openid:
+  enabled: false
+  allowedAudience: ""
+  roleClaim: sub
+    # Set following to true if admin console is configured to use Okta integration. (The admin console uses a local admin token.)
+  attemptAuthenticationProviderToken: true
+  issuerUrl: ""
+  scope: ""
+  client_id: ""
+  client_secret: ""
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/values.yaml:1767]: What is the purpose of this block? I don't see these configs in use.

File: helm-chart-sources/pulsar/values.yaml [hunk 12]
```diff
@@ -1752,6 +1752,31 @@ pulsarHeartbeat:
...
+# For configuring OpenID in Okta for Pulsar, instructions are here: https://www.youtube.com/watch?v=UQBrecHOXxU&ab_channel=DataStaxDevelopers
+# Please also see the docs here: https://github.com/datastax/pulsar-openid-connect-plugin
+# Then, use the values from the Payload generated to populate these fields:
+openid:
+  enabled: false
+  allowedAudience: ""
+  roleClaim: sub
+    # Set following to true if admin console is configured to use Okta integration. (The admin console uses a local admin token.)
+  attemptAuthenticationProviderToken: true
+  issuerUrl: ""
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/values.yaml:1764]: Nit: let's use `issuerUrls` and comment that it is a comma separated list.

File: helm-chart-sources/pulsar/templates/proxy/proxy-ws-configmap.yaml [hunk 13]
```diff
@@ -68,8 +68,30 @@ data:
   superUserRoles: "{{ .Values.superUserRoles }}"
   {{- if .Values.keycloak.enabled }}
   PULSAR_PREFIX_openIDAllowedTokenIssuers: "{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }},{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}.{{ .Release.Namespace }}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }},{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}.{{ template "pulsar.serviceDnsSuffix" . }}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }}"
+  {{- else }}
+  PULSAR_PREFIX_openIDAllowedTokenIssuers: "{{ .Values.openid.issuerUrl }}"
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/proxy/proxy-ws-configmap.yaml:72]: This solution looks good to me. We can document that keycloak uses a slightly different config path, and then eventually we can collapse the logic, if necessary.

File: helm-chart-sources/pulsar/templates/proxy/proxy-configmap.yaml [hunk 14]
```diff
@@ -123,15 +123,51 @@ data:
...
   PULSAR_PREFIX_openIDAllowedTokenIssuers: "{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }},{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}.{{ .Release.Namespace }}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }},{{ template "pulsar.get.http.or.https" . }}{{ template "pulsar.keycloak.fullname" .}}.{{ template "pulsar.serviceDnsSuffix" . }}{{ template "pulsar.keycloak.issuer.port" .}}/realms/{{ .Values.keycloak.realm }}"
+  {{- else }}
+  PULSAR_PREFIX_openIDAllowedTokenIssuers: "{{ .Values.openid.issuerUrl }}"
   {{- end }}
   tokenPublicKey: "file:///pulsar/token-public-key/{{ .Values.tokenPublicKeyFile }}"
   brokerClientAuthenticationPlugin: "org.apache.pulsar.client.impl.auth.AuthenticationToken"
   brokerClientAuthenticationParameters: "file:///pulsar/token-proxy/proxy.jwt"
 {{- end }}
+{{- if .Values.openid.enabled }}
+  PULSAR_PREFIX_openIDAttemptAuthenticationProviderToken: "false"
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/proxy/proxy-configmap.yaml:134]: Looks like we set this below, so we can remove this line.

```suggestion
```

File: helm-chart-sources/pulsar/templates/proxy/proxy-configmap.yaml [hunk 15]
```diff
@@ -123,15 +123,51 @@ data:
...
+  PULSAR_PREFIX_openIDJwkCacheSize: "{{ .Values.openid.jwk.cacheSize }}"
+  PULSAR_PREFIX_openIDJwkExpiresSeconds: "{{ .Values.openid.jwk.expiresSeconds }}"
+  PULSAR_PREFIX_openIDJwkConnectionTimeoutMillis: "{{ .Values.openid.jwk.connectionTimeoutMillis }}"
+  PULSAR_PREFIX_openIDJwkReadTimeoutMillis: "{{ .Values.openid.jwk.readTimeoutMillis }}"
+  PULSAR_PREFIX_openIDMetadataCacheSize: "{{ .Values.openid.metadata.cacheSize }}"
+  PULSAR_PREFIX_openIDMetadataExpiresSeconds: "{{ .Values.openid.metadata.expiresSeconds }}"
+  PULSAR_PREFIX_openIDMetadataConnectionTimeoutMillis: "{{ .Values.openid.metadata.connectionTimeoutMillis }}"
+  PULSAR_PREFIX_openIDMetadataReadTimeoutMillis: "{{ .Values.openid.metadata.readTimeoutMillis }}"
+  PULSAR_PREFIX_openIDRequireHttps: "{{ .Values.openid.requireHttps }}"
+  # Set following to true if admin console is configured to use Okta integration. The admin console uses the user model with local admin token, not Okta integration, according to notes from Ming-Hong Liu.
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/proxy/proxy-configmap.yaml:150]:
```suggestion
  # Set following to true if admin console is configured to use Okta integration. The admin console uses the user model with local admin token, not Okta integration.
```

File: helm-chart-sources/pulsar/templates/proxy/proxy-ws-configmap.yaml [hunk 16]
```diff
@@ -68,8 +68,30 @@ data:
...
+  PULSAR_PREFIX_openIDJwkCacheSize: "{{ .Values.openid.jwk.cacheSize }}"
+  PULSAR_PREFIX_openIDJwkExpiresSeconds: "{{ .Values.openid.jwk.expiresSeconds }}"
+  PULSAR_PREFIX_openIDJwkConnectionTimeoutMillis: "{{ .Values.openid.jwk.connectionTimeoutMillis }}"
+  PULSAR_PREFIX_openIDJwkReadTimeoutMillis: "{{ .Values.openid.jwk.readTimeoutMillis }}"
+  PULSAR_PREFIX_openIDMetadataCacheSize: "{{ .Values.openid.metadata.cacheSize }}"
+  PULSAR_PREFIX_openIDMetadataExpiresSeconds: "{{ .Values.openid.metadata.expiresSeconds }}"
+  PULSAR_PREFIX_openIDMetadataConnectionTimeoutMillis: "{{ .Values.openid.metadata.connectionTimeoutMillis }}"
+  PULSAR_PREFIX_openIDMetadataReadTimeoutMillis: "{{ .Values.openid.metadata.readTimeoutMillis }}"
+  PULSAR_PREFIX_openIDRequireHttps: "{{ .Values.openid.requireHttps }}"
+  # Set following to true if admin console is configured to use Okta integration. The admin console uses the user model with local admin token, not Okta integration, according to notes from Ming-Hong Liu.
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/proxy/proxy-ws-configmap.yaml:92]:
```suggestion
  # Set following to true if admin console is configured to use Okta integration. The admin console uses the user model with local admin token, not Okta integration.
```

File: helm-chart-sources/pulsar/templates/function/function-configmap.yaml [hunk 17]
```diff
@@ -127,6 +127,39 @@ data:
...
+      openIDJwkCacheSize: "{{ .Values.openid.jwk.cacheSize }}"
+      openIDJwkExpiresSeconds: "{{ .Values.openid.jwk.expiresSeconds }}"
+      openIDJwkConnectionTimeoutMillis: "{{ .Values.openid.jwk.connectionTimeoutMillis }}"
+      openIDJwkReadTimeoutMillis: "{{ .Values.openid.jwk.readTimeoutMillis }}"
+      openIDMetadataCacheSize: "{{ .Values.openid.metadata.cacheSize }}"
+      openIDMetadataExpiresSeconds: "{{ .Values.openid.metadata.expiresSeconds }}"
+      openIDMetadataConnectionTimeoutMillis: "{{ .Values.openid.metadata.connectionTimeoutMillis }}"
+      openIDMetadataReadTimeoutMillis: "{{ .Values.openid.metadata.readTimeoutMillis }}"
+      openIDRequireHttps: "{{ .Values.openid.requireHttps }}"
+      # Set following to true if admin console is configured to use Okta integration. The admin console uses the user model with local admin token, not Okta integration, according to notes from Ming-Hong Liu.
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/function/function-configmap.yaml:148]:
```suggestion
      # Set following to true if admin console is configured to use Okta integration. The admin console uses the user model with local admin token, not Okta integration.
```

File: helm-chart-sources/pulsar/values.yaml [hunk 18]
```diff
@@ -1752,6 +1752,31 @@ pulsarHeartbeat:
       topicName: "persistent://public/default/pubsub-latency-test"
   service: {}
 
+# For configuring OpenID in Okta for Pulsar, instructions are here: https://www.youtube.com/watch?v=UQBrecHOXxU&ab_channel=DataStaxDevelopers
+# Please also see the docs here: https://github.com/datastax/pulsar-openid-connect-plugin
+# Then, use the values from the Payload generated to populate these fields:
+openid:
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/values.yaml:1758]: Please add a key value pair here for the `withS4K` defaults. That will give an indication of what is available.

File: helm-chart-sources/pulsar/templates/openid/kafka-configmap.yaml [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+#
+
+{{- if and .Values.openid.enabled .Values.openid.withS4k }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: "openid-s4k-config"
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/openid/kafka-configmap.yaml:22]: I think we try to add the cluster name in the config map name, but I might be misremembering. Will you double check how some of the other config map names are created?

File: examples/kafka/demos/openid_demo.sh [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+
+YOUR_USER="ExampleUser"
+ISSUER_URI="https://dev-42506116.okta.com/oauth2/aus7ypk6sjvgF4l615d7"
+AUDIENCE="api://pulsarClient"
+SCOPE="pulsar_client_m2m"
+AUTH_PARAMS=$(cat <<EOF
+{"privateKey":"/pulsar/conf/creds.json","issuerUrl":"$ISSUER_URI","scope":"$SCOPE","audience":"$AUDIENCE"}
+EOF
+)
+CLIENT_ID="0oa7ypwvxnvo9xnDd5d7"
```
Comment (devinbost) [inline examples/kafka/demos/openid_demo.sh:18]: Reminder to clean up secrets

File: examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+createCertificates:
+  selfSignedPerComponent:
+    enabled: true
+
+openid:
+  enabled: true
+  # From token generated in Okta UI or other method:
+  issuerUrl: https://dev-42506116.okta.com/oauth2/aus7ypk6sjvgF4l615d7
+  scope: pulsar_client_m2m
+  client_id: "0oa7ypwvxnvo9xnDd5d7"
```
Comment (devinbost) [inline examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2-low-resource.yaml:178]: Reminder to replace secrets with placeholders and write in the instructions where to place the secrets and the URL for Okta. In the instructions and demo script, we can do something like helm install --set path.to.secret=SECRET-VALUE …..

File: examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2.yaml [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+createCertificates:
+  selfSignedPerComponent:
+    enabled: true
+
+openid:
+  enabled: true
+  # From token generated in Okta UI or other method:
+  issuerUrl: https://dev-42506116.okta.com/oauth2/aus7ypk6sjvgF4l615d7
+  scope: pulsar_client_m2m
+  client_id: "0oa7ypwvxnvo9xnDd5d7"
```
Comment (devinbost) [inline examples/kafka/dev-values-tls-all-components-and-kafka-and-oauth2.yaml:178]: Reminder to clean up

File: helm-chart-sources/pulsar/values.yaml (hunk 11 above)
Comment (michaeljmarshall) [reply helm-chart-sources/pulsar/values.yaml]: I removed this block, since it isn't used.

File: helm-chart-sources/pulsar/templates/proxy/proxy-configmap.yaml [hunk 23]
```diff
@@ -123,15 +123,42 @@ data:
...
+  PULSAR_PREFIX_openIDMetadataExpiresSeconds: "{{ .Values.openid.metadata.expiresSeconds }}"
+  PULSAR_PREFIX_openIDMetadataConnectionTimeoutMillis: "{{ .Values.openid.metadata.connectionTimeoutMillis }}"
+  PULSAR_PREFIX_openIDMetadataReadTimeoutMillis: "{{ .Values.openid.metadata.readTimeoutMillis }}"
+  PULSAR_PREFIX_openIDRequireHttps: "{{ .Values.openid.requireHttps }}"
+  # Set following to true if admin console is configured to use Okta integration. The admin console uses the user model with local admin token, not Okta integration.
+  PULSAR_PREFIX_openIDAttemptAuthenticationProviderToken: "{{ .Values.openid.attemptAuthenticationProviderToken }}"
+{{- end }}
+{{- if and .Values.openid.enabled .Values.openid.withS4k }}
+  PULSAR_PREFIX_kopOauth2AuthenticateCallbackHandler: "io.streamnative.pulsar.handlers.kop.security.auth.OauthValidatorCallbackHandler"
+  PULSAR_PREFIX_kopOauth2ConfigFile: "/pulsar/conf/kop-handler.properties"
```
Comment (michaeljmarshall) [inline helm-chart-sources/pulsar/templates/proxy/proxy-configmap.yaml:153]: If we're not mounting the secret, we must not need this.

File: README.md [hunk 24]
```diff
@@ -677,3 +678,195 @@ Once you have created the secrets that store the certificate info (or specified
 enableTls: true
 
 ```
+### OpenID with Starlight for Kafka
```
Comment (eolivelli) [inline README.md:681]: Follow up work: move this stuff to a dedicated piece of documentation on the officials docs

File: helm-chart-sources/pulsar/templates/proxy/proxy-configmap.yaml (hunk 23 above)
Comment (eolivelli) [reply helm-chart-sources/pulsar/templates/proxy/proxy-configmap.yaml:153]: Actually this file is static. We could implement a flag in s4k in order to not need to load this file at all

**Review by eolivelli — APPROVED** [review]

**Review by michaeljmarshall — APPROVED** [review]
